src/Sylius/Component/Core/Promotion/Action/PercentageDiscountAction.php

```diff
@@ -11,35 +11,66 @@
 
 namespace Sylius\Component\Core\Promotion\Action;
 
-use Sylius\Component\Core\Model\OrderInterface;
-use Sylius\Component\Core\Model\OrderItemInterface;
-use Sylius\Component\Order\Model\OrderItemUnitInterface;
+use Sylius\Component\Core\Distributor\IntegerDistributorInterface;
+use Sylius\Component\Core\Promotion\Applicator\UnitsPromotionAdjustmentsApplicatorInterface;
+use Sylius\Component\Originator\Originator\OriginatorInterface;
 use Sylius\Component\Promotion\Model\PromotionInterface;
 use Sylius\Component\Promotion\Model\PromotionSubjectInterface;
-use Sylius\Component\Resource\Exception\UnexpectedTypeException;
 
 /**
  * @author Paweł Jędrzejewski <pawel@sylius.org>
  * @author Saša Stamenković <umpirsky@gmail.com>
+ * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
  */
 class PercentageDiscountAction extends DiscountAction
 {
     const TYPE = 'order_percentage_discount';
 
     /**
+     * @var IntegerDistributorInterface
+     */
+    private $distributor;
+
+    /**
+     * @var UnitsPromotionAdjustmentsApplicatorInterface
+     */
+    private $unitsPromotionAdjustmentsApplicator;
+
+    /**
+     * {@inheritdoc}
+     *
+     * @param IntegerDistributorInterface $distributor
+     * @param UnitsPromotionAdjustmentsApplicatorInterface $unitsPromotionAdjustmentsApplicator
+     */
+    public function __construct(
+        OriginatorInterface $originator,
+        IntegerDistributorInterface $distributor,
+        UnitsPromotionAdjustmentsApplicatorInterface $unitsPromotionAdjustmentsApplicator
+    ) {
+        parent::__construct($originator);
+
+        $this->distributor = $distributor;
+        $this->unitsPromotionAdjustmentsApplicator = $unitsPromotionAdjustmentsApplicator;
+    }
+
+    /**
      * {@inheritdoc}
      */
     public function execute(PromotionSubjectInterface $subject, array $configuration, PromotionInterface $promotion)
     {
-        if (!$subject instanceof OrderInterface) {
-            throw new UnexpectedTypeException($subject, OrderInterface::class);
+        if (!$this->isSubjectValid($subject)) {
+            return;
         }
 
-        $adjustment = $this->createAdjustment($promotion);
-        $adjustmentAmount = (int) round($subject->getPromotionSubjectTotal() * $configuration['percentage']);
-        $adjustment->setAmount(-$adjustmentAmount);
+        $this->isConfigurationValid($configuration);
+
+        $promotionAmount = $this->calculateAdjustmentAmount($subject->getPromotionSubjectTotal(), $configuration['percentage']);
+        if (0 === $promotionAmount) {
+            return;
+        }
 
-        $subject->addAdjustment($adjustment);
+        $splitPromotion = $this->distributor->distribute($promotionAmount, $subject->countItems());
```

Comment:
That looks wrong, this should be `->distribute($itemsTotals, $promotionAmount)` as above.